add flag for building bluechi-selinux with any port

engelmi opened the pull request:
Fixes: https://github.com/eclipse-bluechi/bluechi/issues/836

Previously, the SELinux policy for BlueChi only contained one type for the port and was used only by the bluechi-agent. This adds a new port type and and uses one for the bind of bluechi-controller and another for the connect of bluechi-agent.
Based on this, tunable policies are added to turn on/off the option to use any port for both components. 
In addition, it removes the implicit permissive policy so that violations to the policies are enforced by default (can be set permissive later on by the user again).

This PR also updates the RPM spec of BlueChi so that it does not require the semange tool and thus is not pulling in python as a dependency anymore. It adds also another option to the build to enable any port policy for bluechi-controller an bluechi-agent.

It also adds documentation about BueChi's SELinux policy, restructuring the order of the table of contents and introducing the section `Security` which contains the selinux policy and mTLS double proxy approach.

__NOTE:__

The changes for the documentation look like this:
![image](https://github.com/eclipse-bluechi/bluechi/assets/26504678/b5166eba-1554-4368-ae38-480d2873b6e7)

mwperina reviewed (approved):
LGTM

mwperina commented on doc/docs/security/selinux.md:
````diff
@@ -0,0 +1,25 @@
+dnf install bluechi-selinux
+```
+
+By default, the package adds an allow rule for the `bluechi-agent` to connect to BlueChi's default port 842. If the controller port where the agent is connecting to changes, then an AVC entry is created:
````
I'd add some commas for better clarity:

"If the controller port, where the agent is connecting to, changes then an AVC entry is created:"

engelmi replied:
I figured out a way to link to the man page section for the `ControllerPort`. I rephrased and simplified the sentence it a bit.

mwperina replied:
OK

mwperina commented on doc/docs/security/selinux.md:
````diff
@@ -0,0 +1,25 @@
+type=AVC msg=audit(1711635155.532:22): avc:  denied  { name_connect } for  pid=531 comm="bluechi-agent" dest=8450 scontext=system_u:system_r:bluechi_agent_t:s0 tcontext=system_u:object_r:unreserved_port_t:s0 tclass=tcp_socket permissive=1
+```
+
+This rule, however, is not enforcing, but permissive and thus not preventing the `bluechi-agent` to connect to the different host.
````
I'd rephrase:

"However this rule is not enforcing, but permissive, and thus not preventing the `bluechi-agent` to connect to the different host."

engelmi replied:
Changed.

engelmi committed:
added tunable policy to enable any port

Relates to: https://github.com/eclipse-bluechi/bluechi/issues/836

Previously, the SELinux policy for BlueChi only contained one
type for the port and was used only by the bluechi-agent. This
adds a new port type and and uses one for the bind of bluechi-
controller and another for the connect of bluechi-agent.
Based on this, tunable policies are added to turn on/off the
option to use any port for both components.
In addition, it removes the implicit permissive policy so that
violations to the policies are enforced by default (can be set
permissive later on by the user again).

Signed-off-by: Michael Engel <mengel@redhat.com>

engelmi committed:
added option to build SELinux policy with any port enabled

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/836

Based on the previous work, this updates the RPM spec of
BlueChi so that it does not require the semange tool and
thus is not pulling in python as a dependency anymore.
It adds also another option to the build to enable any
port policy for bluechi-controller an bluechi-agent.

Signed-off-by: Michael Engel <mengel@redhat.com>

engelmi committed:
added documentation for BlueChi's revised SELinux policy

This PR reorders the table of contents for the readthedoc
and bundles the mTLS as well as the networking and the
new SELinux subsection together under the Security section.

Signed-off-by: Michael Engel <mengel@redhat.com>

mwperina commented on doc/docs/security/selinux.md:
````diff
@@ -0,0 +1,30 @@
+
+```bash
+# Turn off the any policy for bluechi-controller
+setsebool -P bluechi_port_bind_any 0
````
Wouldn't  `bluechi_controller_bind_any_port` be a more descriptive name?

engelmi replied:
I added the `_controller_` substring, but would keep the `_bind_any` part similar to the [systemd_socket_proxyd_bind_any](https://github.com/fedora-selinux/selinux-policy/blob/85bbb6d1bc69e84ea2440cff647e426b74f1b40a/policy/modules/system/systemd.te#L14)

mwperina replied:
OK

mwperina commented on doc/docs/security/selinux.md:
```diff
@@ -0,0 +1,30 @@
+setsebool -P bluechi_port_bind_any 0
+
+# Turn off the any policy for bluechi-agent
+setsebool -P bluechi_agent_port_connect_any 0
```
Wouldn't be `bluechi_agent_connect_to_any_port` be a more descriptive name?

engelmi replied:
I would keep the `_connect_any` part similar to the [systemd_socket_proxyd_connect_any](https://github.com/fedora-selinux/selinux-policy/blob/85bbb6d1bc69e84ea2440cff647e426b74f1b40a/policy/modules/system/systemd.te#L22C13-L22C46)

mwperina replied:
OK

mwperina commented on selinux/bluechi.te:
```diff
@@ -16,8 +16,11 @@ init_daemon_domain(bluechi_agent_t, bluechi_agent_exec_t)
+corenet_port(bluechi_agent_port_t)
+
+gen_tunable(bluechi_port_bind_any, false)
+gen_tunable(bluechi_agent_port_connect_any, false)
```
Could you please move those boolean options to the top of the file and add a description for them similar as system has?

https://github.com/fedora-selinux/selinux-policy/blob/85bbb6d1bc69e84ea2440cff647e426b74f1b40a/policy/modules/system/systemd.te#L14

engelmi replied:
Done.

mwperina commented on bluechi.spec.in:
```diff
@@ -167,27 +173,21 @@ SELinux policy associated with the bluechi and bluechi-agent daemons
 # Remove hirte policy
 if [ $1 -eq 1 ]; then
-   semanage port -N -d -p udp 842 2>/dev/null || true
-   semanage port -N -d -p tcp 842 2>/dev/null || true
```
Don't we need to keep selinux upgrade from hirte to bluechi till Fedora 39 deprecation? I thought that we would want to add a build option to be able to turn it off on selected platforms?

Also, have you tried upgrade from bluechi 0.7 to bluechi 0.8 with any port policy update? Don't we need to remove those specific port options before installing new policy with any port?

engelmi replied:
These two lines for semanage basically remove the allow rule for tcp and udp on port 842 - no specific domain is given. I don't think these are needed even for the old hirte considering the latest changes. 

> Also, have you tried upgrade from bluechi 0.7 to bluechi 0.8 with any port policy update? Don't we need to remove those specific port options before installing new policy with any port?

Based on how the any port is implemented:
```
tunable_policy(`bluechi_agent_port_connect_any',`
	corenet_tcp_connect_all_ports(bluechi_agent_t)
',`
	allow bluechi_agent_t bluechi_agent_port_t:tcp_socket name_connect;
')
```
Either the labelled port or the any connect/bind is used. Therefore, even though the port labels aren't removed, those don't have any effect till the any is disabled. And even then, the ports would be "pre-configured" - which is ok, I think, considering the new documentation would guide the user to add ports (and thus viewing these). 

I tried verifying this by upgrading from v0.7 to v0.8 with the any port policy and it really kept the port rules:
```bash
[root@localhost tmp]# semanage port -l | grep bluechi
bluechi_port_t                 tcp      842
bluechi_port_t                 udp      842
```
And starting BlueChi on another port was still possible with the any policy active. Switch it off and running the controller on another port, e.g. 8420, was blocked as expected.

mwperina replied:
OK, makes sense

mwperina reviewed (changes requested):
Just noticed that we forgot to remove dependency on policycoreutils-python-utils

rhatdan commented on bluechi.spec.in:
```diff
@@ -167,27 +171,21 @@ SELinux policy associated with the bluechi and bluechi-agent daemons
-semanage port -a -t bluechi_port_t -p udp 842 2>/dev/null || semanage port -m -t bluechi_port_t -p udp 842
-semanage port -a -t bluechi_port_t -p tcp 842 2>/dev/null || semanage port -m -t bluechi_port_t -p tcp 842
+%if %{with_selinux_port_any}
+   setsebool -P bluechi_controller_port_bind_any 1
```
combine these lines together, to save install time.

  setsebool -P bluechi_controller_port_bind_any=1 bluechi_agent_port_connect_any=1

engelmi replied:
Good point, changed it.

rhatdan commented on selinux/bluechi.te:
```diff
@@ -5,6 +5,12 @@ policy_module(bluechi, 1.0.0)
 #
 
+# Allow bluechi_t to bind to any port instead of one labelled with bluechi_port_t.
+gen_tunable(bluechi_controller_port_bind_any, false)
```
Why not change the default on these to true and then allow users to turn up the security rather then forcing everyone to turn down the security.

alexlarsson replied:
Yeah, i think it makes sense to do this, otherwise the default isn't gonna work. If this defaults to true, then only in the case where you want to limit to a port do you have to change anything (setting both any to false and the port).

engelmi replied:
Good point! I changed both anys to default to `true`.

mwperina commented on README.developer.md:
```diff
@@ -377,9 +377,12 @@ The package build process can be adjusted using following environment variables:
 - `WITH_COVERAGE`
-  - To start collecting coverage this option should contain `1`.
+  - To start collecting coverage this option should contain `1`. Default: `0`.
+- `WITH_SELINUX_PORT_ANY`
```
Do we need that build option when both `bluechi_controller_port_bind_any` and `bluechi_agent_port_connect_any` are enabled by default? Do we really need to rebuild the package when we want to limit access to specific port only? Can't limitation to specific port be achieved only after installation by turning off those selinux booleans and run semanage?

engelmi replied:
Good point! I removed it and verified that the default in the `bluechi.te` works as expected. Removed the `with_selinux_port_any` option from the spec file.

mwperina commented on doc/docs/security/selinux.md:
````diff
@@ -0,0 +1,44 @@
+semanage port -a -t bluechi_agent_port_t -p tcp <port>
+```
+
+## Change from enforcing to permissive
````
I think we are trying to avoid permissive mode in general, so is it worth to have that section?

engelmi replied:
I think its not bad to have it. Its not much effort and if somebody for some reason really wants to switch it to permissive, it briefly described here.
